src/transforms/sample/transform.rs

```diff
@@ -104,19 +172,14 @@ impl FunctionTransform for Sample {
             Event::Metric(_) => panic!("component can never receive metric events"),
         });
 
-        let counter_value: u64 = *self.counter.entry(group_by_key.clone()).or_default();
-
-        let num = if let Some(value) = value {
-            seahash::hash(value.as_bytes())
+        let threshold_exceeded = self.rate.increment(&group_by_key);
+        let should_process = if let Some(value) = value {
+            self.rate.hash_within_ratio(value.as_bytes())
         } else {
-            counter_value
+            threshold_exceeded
         };
```

Comment:
Since `threshold_exceeded` is only used if `value.is_none()`, I would suggest to move this whole condition inside that `increment` function, which will likely suggest even more simplifications.